File: haskell/fib.hs
import System.Environment
import Data.List (unfoldr)
import Text.Printf

fib :: Integer -> Integer
fib n = if n < 2 then n else fib (n - 1) + fib (n - 2)

fibPattern :: Integer -> Integer
fibPattern n
  | n < 2 = n
  | otherwise = fibPattern (n - 1) + fibPattern (n - 2)

fibTailRecursive :: Integer -> Integer
fibTailRecursive n = fibTailAuxiliary n 0 1
  where fibTailAuxiliary :: Integer -> Integer -> Integer -> Integer
        fibTailAuxiliary 0 prevFib _ = prevFib
        fibTailAuxiliary n prevFib fib =
          fibTailAuxiliary (n - 1) fib (prevFib + fib)

fibFormula :: Double -> Integer
fibFormula n = round (((5 ** 0.5 + 1) / 2) ** n / 5 ** 0.5)

fibFastDoubling :: Integer -> Integer
fibFastDoubling n = fst (fibFastAuxiliary n)
  where fibFastAuxiliary :: Integer -> (Integer, Integer)
        fibFastAuxiliary 0 = (0, 1)
        fibFastAuxiliary n = do
            let (prevFib, fib) = fibFastAuxiliary (n `div` 2)
            let a = prevFib * (fib * 2 - prevFib)
            let b = prevFib * prevFib + fib * fib
            if n `mod` 2 == 0
              then (a, b)
              else (b, a + b)

fibFormal :: Integer -> Integer
fibFormal 0 = 0
fibFormal 1 = 1
fibFormal n = fibFormal (n - 1) + fibFormal (n - 2)

-- Callable by `fibUnfold!!input`
fibUnfold :: [Integer]
fibUnfold = unfoldr (\(prevFib, fib) -> Just (prevFib, (fib, prevFib + fib))) (0, 1)

fibZip :: [Integer]
fibZip = 0 : 1 : zipWith (+) fibZip (tail fibZip)

fibBinet :: Integer -> Integer
fibBinet n = round $ phi ** fromIntegral n / sq5
  where
    sq5 = sqrt 5 :: Double
    phi = (1 + sq5) / 2

main :: IO ()
main = do
  args <- getArgs
  let input = if length args > 0 then read $ head args else 29
  printf "Fibonacci %d: %d\n" input (fibFastDoubling input)

File: lua/fib.lua
function fib(n)
  return n < 2 and n or fib(n - 1) + fib(n - 2)
end

function fibLinear(n)
  if n < 2 then return n end
  local prevFib = 0
  local fib = 1
  while n > 1 do
    prevFib, fib = fib, prevFib + fib
    n = n - 1
  end
  return fib
end

function round(float)
  local int, part = math.modf(float)
  return float == math.abs(float) and part >= .5 and int+1 or part <= -.5 and int-1 or int
end
function fibFormula(n)
  return round(((5 ^ .5 + 1) / 2) ^ n / 5 ^ .5)
end

function fibTailAuxiliary(n, prevFib, fib)
  return n == 0 and prevFib or fibTailAuxiliary(n - 1, fib, prevFib + fib)
end
function fibTailRecursive(n, prevFib, fib)
  return fibTailAuxiliary(n, 0, 1)
end

function fibFastAuxiliary(n)
  if n == 0 then return {0, 1} end
  local result = fibFastAuxiliary(math.floor(n / 2))
  local prevFib, fib = result[1], result[2]
  local c = prevFib * (fib * 2 - prevFib)
  local d = prevFib * prevFib + fib * fib
  return n % 2 == 0 and {c, d} or {d, c + d}
end
function fibFastDoubling(n)
  return fibFastAuxiliary(n)[1]
end

input = #arg > 0 and tonumber(arg[1]) or 29
print(string.format("Fibonacci %d: %d", input, fibFastDoubling(input)))
